### ISSUE TITLE
Scheduling / simultanious actions on same server

### ISSUE THREAD

**Issue by jjnxpct** [akeeba/panopticon#673]
### Due diligence

- [X] I have searched all open and closed issues for this problem.
- [X] I have searched all open and closed discussions for this problem.
- [X] I have looked at the CHANGELOG to make sure this problem has not already been addressed.
- [X] I agree to the Code of Conduct of this repository

### Describe the issue

Hi! I think Panopticon is a interesting development. I am planning to test it a see if it fits our needs. Thanks for starting this service.

I do have a general concern: When we host Panoptiocon on our own server and add about 100 sites to it that are also on the same server, is there a mechanism that prevent starting to many client side process (files scan,. backup update) at the same time?

We should probably schedule backups at different times / timeslots. And also the file scanner. Correct?

But when we enable autoupdates fo an extension (maybe like Akeeba Admin Tools) on all sites, will Panopticon know not to start this update on all sites at once? That would cause serious issues on our server. Same goes for installing or updating extensions on all site (is that a feature at this point?) 

So basically I would like to be sure that tasks that we can not plan ourselfs will not overload our server, because all our sites run on the same server.

### Panopticon version

latest

### PHP version

8.3.6

### Database type

MariaDB

### Database version

not sure

### Browser

None

**Comment by nikosdion** [comment]
There is no good, reliable, universal way of knowing a. which sites are on the same server and b. what is the maximum load the server can bear. This is something _you_ know, as the systems admin.

Therefore, yes, it's up to you to set up your backup, file scanner, and updates schedule so that you don't cause unsustainable demand on the server resources. This is why you get these options.

As for extension updates, remember that it won't be 100 sites updating at the same time. The number of sites being updated will **at most** be three times the number of Panopticon CRON jobs you have. If you have three CRON jobs, **at worst** you'll have nine sites updating at the same time. If your server can't handle that, I doubt it can realistically handle 100 sites being hosted on it ;)

**Comment by brianteeman** [comment]
Should also add that running panopticon on the exact same server that you are hosting the sites on is not the best idea.

**Comment by nikosdion** [comment]
Please read https://github.com/akeeba/panopticon/wiki/CRON-Jobs

The more CRON jobs you run, the more capacity for scheduled tasks you add. Remember, PHP can do _one thing at a time_. Having multiple CRON jobs running in parallel means you have multiple PHP thread each doing one thing at a time, therefore you have this many things happening at a time. 

As to why this may be multiplied by three, read the info on the `--loop` option.

**Comment by jjnxpct** [comment]
Ah, OK. That helps, a bit, But  I'm not sure I am there yet.... It's still a bit confusing to me. Sorry. Hope you don't mind some more clarification:

All actions (for all added sites) in Panopticon (scheduled backups, scheduled file scan, extension auto updates, and also all manual actions started from within Panopticon) are added to the tasks queue that is then 'processed' by the the cron script every minute. Correct?

Does the cron script only execute one task at the time before moving on to the next task in the queue?  Does the next task only get triggered when a previous task send to a website returns an 'OK' of 'Finished'? 

_"This is where the --loop option comes in. When this option is present, the script processes all pending tasks. "_ 

At once or sequential?

Adding more the one cron will be a way to trigger more then one task at the same time so more tasks can be processed at once? As long as the server can handle this properly we could potentially add 10 or more cronjobs?

**Comment by nikosdion** [comment]
> All actions (for all added sites) in Panopticon (scheduled backups, scheduled file scan, extension auto updates, and also all manual actions started from within Panopticon) are added to the tasks queue that is then 'processed' by the the cron script every minute.

Correct.

> Does the cron script only execute one task at the time before moving on to the next task in the queue? Does the next task only get triggered when a previous task send to a website returns an 'OK' of 'Finished'?

Correct. As I said, PHP can do **one thing at a time**. It does not do multithreaded execution.

> At once or sequential?

Always sequential. PHP can only do one thing at once.

Do not that --loop is about when the CRON job terminates. Using --loop uses the maximum amount of time possible. This means that one CRON job may be running for longer than a minute. Therefore, its next instance will be created before its previous one, or even two previous ones have exited. This is where the multiplier comes from. Each instance runs in parallel with the others, it is completely independent of other instances of the same or different CRON jobs, and has no idea about the existence of other instances.

> Adding more the one cron will be a way to trigger more then one task at the same time so more tasks can be processed at once? 

This is exactly what I have written in the page I told you to read.

> As long as the server can handle this properly we could potentially add 10 or more cronjobs?

Exactly so. Each CRON job is a task execution pipeline which can execute one task at any given time. The more CRON jobs you have, the more task execution pipelines you add.


OK, here's an imperfect analogy.

Think of CRON jobs as CPU cores and --loop as hyperthreading.

Back in the olden days, CPUs had one core and no hyperthreading. They could do one thing at a time. That's what you get with one CRON job.

Then, they got hyperthreading. A CPU core could do approximately two things at once if the timing was right. That's what you get with one CRON job and --loop.

Then, CPUs got multiple cores and hyperthreading. A quadcore CPU could do four to eight things at once. That's what you get with four CRON jobs and --loop.

A CPU won't get any faster if there is no load to process, or if it gets a deluge of load at a short period of time exceeding its processing capability. The same goes with the CRON jobs. Just because you have 10 CRON jobs doesn't mean everything gets 10 times faster if there is not enough to work, and definitely not if you try to let's say run 100 backups at the same time (you'd still get only 10 to 30 backups running in parallel if you used --loop).

Just because a CPU has multiple cores doesn't mean that all load will be sped up. Some load, such as finding prime numbers within a range, will only ever run on a single core since the output of the previous step is used to run the next step of the calculation. Likewise, having 10 CRON jobs won't mean that backups, or core/extension updates will get 10 times faster. The steps for these tasks need to run sequentially, each one starting after the previous one finishes up.

**Comment by jjnxpct** [comment]
OK, now I understand. I took me some time but I'm glad you stuck with me and helped clarify this. Thanks!

**Comment by nikosdion** [comment]
You're welcome!
